File: python/131-140/Clone Graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    dict = {}
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if not node:
            return None
        
        if node not in self.dict:
            self.dict[node] = UndirectedGraphNode(node.label)
            for e in node.neighbors:
                self.dict[node].neighbors.append(self.cloneGraph(e))

    return self.dict[node]

File: cpp/131-140/Clone Graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;

    Node() {}

    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/
class Solution {
    unordered_map<Node*, Node*> dict;
public:
    
    Node* cloneGraph(Node* node) {
        //if (!node) return nullptr;
        if (dict.count(node) == 0) {
            dict[node] = new Node();
            dict[node]->val = node->val;
            for (Node* neighbor : node->neighbors) {
                dict[node]->neighbors.push_back(cloneGraph(neighbor));
            }
        }
        return dict[node];
    }
};
